url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tf/Partition.java

```diff
@@ -0,0 +1,199 @@
+package uy.edu.fing.mina.fsa.tf;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import uy.edu.fing.mina.fsa.utils.CombinationGenerator;
+
+public class Partition {
+
+	public List<TfI> left = null;
+	public List<TfI> right = null;
+	
+	public Partition(List<TfI> left, List<TfI> right) {
+		super();
+		this.left = left;
+		this.right = right;
+	}
+
+	@Override
+	public String toString() {
+		return left + "-" + right;
+	}
+
+  /**
+   * @param relation
+   * @return
+   */
+  public static TfI toTfrelation3(Partition relation) {
+  
+    TfI tfrelation = null;
+  
+    if (relation.left != null) {
+      if (relation.left.size() == 1) 
+    	System.out.println(relation.left);
+      for (Iterator<TfI> iter = relation.left.iterator(); iter.hasNext();) {
+        TfI element = iter.next();
+        if (tfrelation == null) tfrelation = element;
+        else
+          tfrelation = tfrelation.andSimple(element);
+      }
+    }
+    if (relation.right != null) {
+      for (Iterator<TfI> iter2 = relation.right.iterator(); iter2.hasNext();) {
+        TfI element = iter2.next();
+        tfrelation = tfrelation.andSimple(element.not());
+      }
+    }
+    return tfrelation;
+  }
+
+  /**
+    * @param partitions
+    * @return
+    */
+   public static Set<TfPair> cleanPartitions(Set<TfPair> partitions) {
+      
+      for (Iterator<TfPair> iter = partitions.iterator(); iter.hasNext();) {
+         TfPair partition = (TfPair) iter.next();
+  
+         TfI tfpartition = null;
+         if (partition.getTfIn() != null && partition.getTfOut() != null)
+            tfpartition = partition.getTfIn().and(partition.getTfOut().not());
+         else if (partition.getTfIn() != null ) 
+            tfpartition = partition.getTfIn();
+         else if (partition.getTfOut() != null) 
+            tfpartition = partition.getTfOut().not();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 64:69, 81:86.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.